internal/github/contributions.go

```diff
@@ -20,53 +25,133 @@ type Contributions struct {
 	Days               []Day
 }
 
-func (g *GithubService) GetContributionsByUsername(ctx context.Context, username string) (*Contributions, error) {
-	var contributionsQuery struct {
-		User struct {
-			ContributionsCollection struct {
-				ContributionCalendar struct {
-					TotalContributions int
-					Weeks              []struct {
-						ContributionDays []struct {
-							ContributionCount int
-							Weekday           int
-							Date              string
-						}
-					}
-				}
-			}
-		} `graphql:"user(login: $username)"`
+type GetContributionsByUsernameOptions struct {
+	Username string
+	From time.Time
+	To time.Time
+}
+
+func (g *GithubService) GetContributionsByUsername(ctx context.Context, options GetContributionsByUsernameOptions) (*Contributions, error) {
+	var contributions *Contributions
+
+	if len(options.Username) == 0 {
+		return nil, ErrMissingUsername
 	}
 
-	err := g.githubClient.Query(ctx, &contributionsQuery, map[string]interface{}{
-		"username": githubv4.String(username),
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "github client query")
+	from, to := options.From, options.To
+
+	if to.IsZero() {
+		to = time.Now()
+	}
+
+	if from.IsZero() {
+		from = to.AddDate(-1, 0, 0)
 	}
 
-	contributions := &Contributions{
-		TotalContributions: contributionsQuery.User.ContributionsCollection.ContributionCalendar.TotalContributions,
+	years := date.GetYearsBetweenDates(from, to)
+
+	contributions = &Contributions{
+		TotalContributions: 			0,
 		Days:               []Day{},
 	}
 
-	for _, w := range contributionsQuery.User.ContributionsCollection.ContributionCalendar.Weeks {
+	if len(years) > 1 {
```

Comment:
I think there is a simplification here where we don't need an if/else
We should be able to execute a loop for both the single and multiple query scenarios and the loop can drive the continue/exit logic

Reply:
@adavila0703 something alone these lines can prevent if/else
```go
	originalFrom := from

	for {

		from = to.AddDate(-1, 0, 0)
		if from.Before(originalFrom) {
			from = originalFrom
		}

		var contributionsQuery struct {
			User struct {
				ContributionsCollection struct {
					ContributionCalendar struct {
						TotalContributions int
						Weeks              []struct {
							ContributionDays []struct {
								ContributionCount int
								Weekday           int
								Date              string
							}
						}
					}
				} `graphql:"contributionsCollection(from: $from, to: $to)"`
			} `graphql:"user(login: $username)"`
		}

		err := g.githubClient.Query(ctx, &contributionsQuery, map[string]interface{}{
			"username": githubv4.String(options.Username),
			"from":     githubv4.DateTime{Time: from},
			"to":       githubv4.DateTime{Time: to},
		})
		if err != nil {
			return nil, errors.Wrap(err, "github client query")
		}

		contributions.TotalContributions += contributionsQuery.User.ContributionsCollection.ContributionCalendar.TotalContributions

		for _, w := range contributionsQuery.User.ContributionsCollection.ContributionCalendar.Weeks {

			for _, d := range w.ContributionDays {

				date, err := time.Parse("2006-01-02", d.Date)
				if err != nil {
					return nil, errors.Wrap(err, "parsing date")
				}

				contributions.Days = append(contributions.Days, Day{
					ContributionCount: d.ContributionCount,
					Weekday:           d.Weekday,
					Date:              date,
				})
			}
		}

		to = from.AddDate(0, 0, -1)

		if from.Equal(originalFrom) {
			break
		}

	}
```